### PR TITLE
Refactor data record merge at pipeline

### PR DESCRIPTION
**Pull request by azexcy**
Changes proposed in this pull request:
  - Refactor data record merge at pipeline
  - Add more E2E log
  - Fix E2E error because of DataRecord hashcode generate method.

---

Before committing this PR, I'm sure that I have checked the following options:
- [ ] My code follows the [code of conduct](https://shardingsphere.apache.org/community/en/involved/conduct/code/) of this project.
- [ ] I have self-reviewed the commit code.
- [ ] I have (or in comment I request) added corresponding labels for the pull request.
- [ ] I have passed maven check locally : `./mvnw clean install -B -T1C -Dmaven.javadoc.skip -Dmaven.jacoco.skip -e`.
- [ ] I have made corresponding changes to the documentation.
- [ ] I have added corresponding unit tests for my changes.

### CODE REVIEW

File: kernel/data-pipeline/core/src/main/java/org/apache/shardingsphere/data/pipeline/core/importer/DataRecordMerger.java [hunk 1]
```diff
@@ -77,89 +60,11 @@ public List<DataRecord> merge(final List<DataRecord> dataRecords) {
      */
     public List<GroupedDataRecord> group(final List<DataRecord> dataRecords) {
         List<GroupedDataRecord> result = new ArrayList<>(100);
-        List<DataRecord> mergedDataRecords = dataRecords.get(0).getUniqueKeyValue().isEmpty() ? dataRecords : merge(dataRecords);
```
Comment (sandynz) [inline kernel/data-pipeline/core/src/main/java/org/apache/shardingsphere/data/pipeline/core/importer/DataRecordMerger.java:80]: Looks `dataRecords.get(0).getUniqueKeyValue().isEmpty()` is ignored

File: kernel/data-pipeline/core/src/main/java/org/apache/shardingsphere/data/pipeline/core/importer/DataSourceImporter.java [hunk 2]
```diff
@@ -242,7 +256,11 @@ private void executeBatchDelete(final Connection connection, final List<DataReco
             for (DataRecord each : dataRecords) {
                 conditionColumns = RecordUtils.extractConditionColumns(each, importerConfig.getShardingColumns(each.getTableName()));
                 for (int i = 0; i < conditionColumns.size(); i++) {
-                    preparedStatement.setObject(i + 1, conditionColumns.get(i).getOldValue());
+                    Object oldValue = conditionColumns.get(i).getOldValue();
+                    if (null == oldValue) {
+                        log.warn("Old value is null, column={}", conditionColumns.get(i));
+                    }
```
Comment (sandynz) [inline kernel/data-pipeline/core/src/main/java/org/apache/shardingsphere/data/pipeline/core/importer/DataSourceImporter.java:262]: Could we print `DataRecord each`? For better troubleshooting

**Review by sandynz — CHANGES_REQUESTED** [review]

File: kernel/data-pipeline/core/src/main/java/org/apache/shardingsphere/data/pipeline/core/importer/DataSourceImporter.java [hunk 3]
```diff
@@ -258,18 +242,46 @@ private void executeBatchDelete(final Connection connection, final List<DataReco
...
                     }
                     preparedStatement.setObject(i + 1, oldValue);
                 }
                 preparedStatement.addBatch();
             }
-            preparedStatement.executeBatch();
+            int[] executeBatch = preparedStatement.executeBatch();
+            if (executeBatch.length == 0) {
+                log.warn("executeDelete failed, deleteSQL={}, conditionColumns={}", deleteSQL, conditionColumns);
+            }
```
Comment (sandynz) [inline kernel/data-pipeline/core/src/main/java/org/apache/shardingsphere/data/pipeline/core/importer/DataSourceImporter.java:255]: Need to verify every execution result, but not `executeBatch.length == 0`

File: kernel/data-pipeline/core/src/main/java/org/apache/shardingsphere/data/pipeline/core/importer/DataSourceImporter.java [hunk 4]
```diff
@@ -258,18 +242,46 @@ private void executeBatchDelete(final Connection connection, final List<DataReco
...
             batchDeleteStatement = null;
         }
     }
     
+    private void sequentialFlush(final DataSource dataSource, final List<DataRecord> buffer) {
+        try (Connection connection = dataSource.getConnection()) {
+            // TODO it's better use transaction, but execute delete maybe not effect when open transaction of PostgreSQL
+            if (null != rateLimitAlgorithm) {
+                rateLimitAlgorithm.intercept(JobOperationType.INSERT, 1);
+            }
```
Comment (sandynz) [inline kernel/data-pipeline/core/src/main/java/org/apache/shardingsphere/data/pipeline/core/importer/DataSourceImporter.java:266]: Could not use hard-coded `JobOperationType.INSERT`

File: kernel/data-pipeline/core/src/main/java/org/apache/shardingsphere/data/pipeline/core/importer/DataSourceImporter.java [hunk 5]
```diff
@@ -258,18 +242,46 @@ private void executeBatchDelete(final Connection connection, final List<DataReco
...
+            int[] executeBatch = preparedStatement.executeBatch();
+            if (executeBatch.length == 0) {
+                log.warn("executeDelete failed, deleteSQL={}, conditionColumns={}", deleteSQL, conditionColumns);
+            }
         } finally {
             batchDeleteStatement = null;
         }
     }
     
+    private void sequentialFlush(final DataSource dataSource, final List<DataRecord> buffer) {
```
Comment (sandynz) [inline kernel/data-pipeline/core/src/main/java/org/apache/shardingsphere/data/pipeline/core/importer/DataSourceImporter.java:261]: If there's no unique key, does it run through `sequentialFlush`?

File: kernel/data-pipeline/core/src/main/java/org/apache/shardingsphere/data/pipeline/core/importer/DataSourceImporter.java [hunk 6]
```diff
@@ -258,18 +242,46 @@ private void executeBatchDelete(final Connection connection, final List<DataReco
...
+                        break;
+                    case IngestDataChangeType.DELETE:
+                        executeBatchDelete(connection, Collections.singletonList(each));
+                        break;
+                    default:
+                }
+            }
+        } catch (final SQLException ex) {
+            throw new PipelineImporterJobWriteException(ex);
+        }
```
Comment (sandynz) [inline kernel/data-pipeline/core/src/main/java/org/apache/shardingsphere/data/pipeline/core/importer/DataSourceImporter.java:283]: It's better to add DataRecord in exception for better troubleshooting

**Review by sandynz — CHANGES_REQUESTED** [review]

File: kernel/data-pipeline/core/src/main/java/org/apache/shardingsphere/data/pipeline/core/importer/DataSourceImporter.java (hunk 5 above)
Comment (azexcy) [reply kernel/data-pipeline/core/src/main/java/org/apache/shardingsphere/data/pipeline/core/importer/DataSourceImporter.java:261]: I find unique key update, delete event not handle correctly now, i will fix it at next PR.

**Review by sandynz — APPROVED** [review]

**Pull request merged**